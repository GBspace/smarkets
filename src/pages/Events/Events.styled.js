import { styled } from "@mui/system";

export const StyledEvents = styled('div')(({theme}) => ({
    display: 'flex',
    flexDirection: 'column',
    alignSelf: 'flex-start',
    padding: '10px',
  }));