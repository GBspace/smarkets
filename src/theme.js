import {createTheme} from '@mui/material/styles';
import {red} from '@material-ui/core/colors';

// A custom theme for this app
const customTheme = createTheme({
    // palette: {
    //   type: 'light',
    //   primary: {
    //     main: '#696969',
    //     contrastText: 'white'
    //   },
    //   secondary: {
    //     main: '#19857b',
    //   },
    //   error: {
    //     main: red.A400,
    //   },
    //   background: {
    //     default: '#282c34',
    //   },
    //   action:{
    //       focus: '#e6e6e6'
    //   }
    // },
    overrides: {
      MuiPaper: {
        root: {
          padding: '20px 10px',
          margin: '10px',
          backgroundColor: '#fff',
        },
      },
      MuiButton: {
        root: {
          margin: '5px',
        },
      },
    },
  });
  export {customTheme};