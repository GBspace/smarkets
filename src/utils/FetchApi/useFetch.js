import { useEffect, useState, useRef } from "react";
  const useFetch = (url, ref, initialValue) => {
    const [data, setData] = useState(initialValue);
    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(true);
    var cors_api_url = 'https://cors-anywhere.herokuapp.com/';
    useEffect(() => {
      if(ref.current){
        fetch(cors_api_url + url, {
          method: 'GET',
          headers: {'Content-Type': 'application/x-www-form-urlencoded'}
        })
        .then(res => {
          if(res.ok) {
            return res.json();
          } else {
            throw new Error('Something went wrong');
          }
        })
        .then(json => { 
          setData(json)
        })
        .catch(err => {
          setError(err.message);
        }).finally(()=>setLoading(false));
      }
      return () => {
        ref.current = false;
      }
    }, [url,ref]);
    return {data,error,loading};
  }

export {useFetch};