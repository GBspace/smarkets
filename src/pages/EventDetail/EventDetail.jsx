import React, {useRef} from 'react'
import { useParams } from 'react-router-dom'
import {useFetch} from '../../utils/FetchApi/useFetch';

const EventDetail = () => {
    const isComponentMounted = useRef(true);
    const {id} = useParams();
    const {data,error,loading} = useFetch(`https://api.smarkets.com/v3/events/${id}/`,isComponentMounted,[]);
    return (
        <div data-testid="eventDetail">
            {loading && <div className='loading'>Loading...</div>}
            {(data?.events) && 
                    <div className='data-container'>
                        <div className='name'>Event Name: {data.events[0].name} </div>
                        <div className='desc'>Event Description: {data.events[0].description} </div>
                        <div className='startTime'>Start Time: {data.events[0].start_datetime} </div>
                    </div>
            }
            {error && <div className="error-container">{error}</div>}
        </div>
    )
}

export default EventDetail;