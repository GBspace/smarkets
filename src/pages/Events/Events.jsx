import { List } from '@material-ui/core';
import React, { useRef } from 'react'
import Event from '../../components/Event/Event';
import {useFetch} from '../../utils/FetchApi/useFetch';
import { StyledEvents } from './Events.styled';

const Events = () => {
  const isComponentMounted = useRef(true);
  const {data,error,loading} = useFetch('https://api.smarkets.com/v3/popular/event_ids/sport/football/',isComponentMounted,[]);
  const events = (data && data.popular_event_ids) || [];
    return ( 
      <StyledEvents>
        <div className='events-container' data-testid="events">
          {loading &&  <div className='loading'>Loading...</div>}
          {data && 
          <List>
            {(events.map(id => <Event key={id} id={id}/>))}
          </List>
          }
          {error && <div className='error-container'>{error}</div>}
        </div>
      </StyledEvents>
  )
}

export default Events;