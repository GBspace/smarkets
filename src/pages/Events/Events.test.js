import Events from './Events';
import EventDetail from '../EventDetail/EventDetail';
import * as React from "react";
import {shallow} from 'enzyme';
import  {useFetch} from '../../utils/FetchApi/useFetch';
import {customTheme} from '../../theme';
import toJson from 'enzyme-to-json';

  jest.mock('../../utils/FetchApi/useFetch', ()=>{
    const spy = jest.fn();
    return {
        __esModule: true,
    ...jest.requireActual('../../utils/FetchApi/useFetch'),
    useFetch: spy,
    }
});

describe('EventsPage', ()=>{
    it('renders correctly when events are fetched', () => {
        useFetch.mockImplementationOnce(()=>({
            data: {
                popular_event_ids: [1,2,3]
            }, 
            loading: false,
            error: null
        }));
        const wrapper = shallow(<Events theme={customTheme}/>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should display loading till data is fetched', ()=>{
        useFetch.mockImplementationOnce(()=>({
            data: null, 
            loading: true,
            error: null
        }));
        const wrapper = shallow(<Events/>);
        const loadingDiv = wrapper.find('.loading');
        expect(loadingDiv.text()).toBe("Loading...");
        });

    it('should display the events', ()=>{
        useFetch.mockImplementationOnce(()=>({
            data: {
                popular_event_ids: [1,2,3]
            }, 
            loading: false,
            error: null
        }));
        const wrapper = shallow(<Events/>);
        const loadingDiv = wrapper.find('.loading');
        const eventComponent = wrapper.children().find('Event');
        expect(eventComponent).toHaveLength(3);
        expect((loadingDiv).exists()).toBe(false);
    });

    it('should display the message in case of no events are scheduled', ()=>{
        useFetch.mockImplementationOnce(()=>({
            data: {
                // popular_event_ids: []
            }, 
            loading: false,
            error: null
        }));
        const wrapper = shallow(<Events/>);
        const loadingDiv = wrapper.find('.loading');
        const eventComponent = wrapper.children().find('Event');
        expect(eventComponent).toHaveLength(0);
        expect((loadingDiv).exists()).toBe(false);
    });

    it('should display Error with a failing newwork request', ()=>{
        useFetch.mockImplementationOnce(()=>({
            data: null, 
            loading: false,
            error: 'Something went wrong'
        }));
        const wrapper = shallow(<Events/>);
        const errorDiv = wrapper.find('.error-container');
        expect(errorDiv.text()).toBe("Something went wrong");
    });
})