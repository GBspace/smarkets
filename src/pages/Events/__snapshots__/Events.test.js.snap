// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsPage renders correctly when events are fetched 1`] = `
<Styled(div)>
  <div
    className="events-container"
    data-testid="events"
  >
    <WithStyles(ForwardRef(List))>
      <Event
        id={1}
        key="1"
      />
      <Event
        id={2}
        key="2"
      />
      <Event
        id={3}
        key="3"
      />
      <Event
        id={4}
        key="4"
      />
    </WithStyles(ForwardRef(List))>
  </div>
</Styled(div)>
`;
